File: gui/src/main/java/io/github/manami/gui/utility/ReadOnlyObservableValue.java
package io.github.manami.gui.utility;

import javafx.beans.InvalidationListener;
import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;

public abstract class ReadOnlyObservableValue<T> implements ObservableValue<T> {

    @Override
    public void addListener(ChangeListener<? super T> listener) {
    }

    @Override
    public void removeListener(ChangeListener<? super T> listener) {
    }

    @Override
    public abstract T getValue();

    @Override
    public void addListener(InvalidationListener listener) {
    }

    @Override
    public void removeListener(InvalidationListener listener) {
    }
}

File: manami-gui/src/main/kotlin/io/github/manamiproject/manami/gui/ReadOnlyObservableValue.kt
package io.github.manamiproject.manami.gui

import javafx.beans.InvalidationListener
import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableValue

data class ReadOnlyObservableValue<T>(val obj: T) : ObservableValue<T> {

    constructor(creator: () -> T): this(creator())

    override fun addListener(listener: ChangeListener<in T>?) {
    }

    override fun addListener(listener: InvalidationListener?) {
    }

    override fun removeListener(listener: InvalidationListener?) {
    }

    override fun removeListener(listener: ChangeListener<in T>?) {
    }

    override fun getValue(): T = obj
}